src/image-tvg.c

```diff
@@ -0,0 +1,1305 @@
+/* get the path command index/id */
+#define TVG_PATH_CMD_INDEX(x) (x & 0x7)
+/* flag indicating the path has line/stroke to it */
+#define TVG_PATH_CMD_HAS_LINE(x) ((x >> 4) & 0x1)
+/* flag indicating the arc is a large arc */
+#define TVG_ARC_LARGE(x) (x & 0x1)
+/* flag indicating the sweep direction 0=left, 1=right */
+#define TVG_ARC_SWEEP(x) ((x >> 1) & 1)
+
+/* F32 pixel format */
+typedef struct {
+    float r, g, b, a;
+} tvg_f32_pixel_t;
+
+/* rgba32 color struct */
+typedef struct {
+    uint8_t r, g, b, a;
+} tvg_rgba32_t;
+
+/* TVG internal color struct */
+typedef struct {
+    float r, g, b;
+} tvg_rgb_t;
+
+/* TVG internal color struct with alpha channel */
+typedef struct {
+    float r, g, b, a;
+} tvg_rgba_t;
+
+/* coordinate */
+typedef struct {
+    float x, y;
+} tvg_point_t;
+
+/* rectangle */
+typedef struct {
+    float x, y;
+    float width, height;
+} tvg_rect_t;
+
+/* gradient data */
+typedef struct {
+    tvg_point_t point0;
+    tvg_point_t point1;
+    uint32_t color0;
+    uint32_t color1;
+} tvg_gradient_t;
+
+/* style data */
+typedef struct {
+    uint8_t kind;
+    union {
+        uint32_t flat;
+        tvg_gradient_t linear;
+        tvg_gradient_t radial;
+    };
+} tvg_style_t;
+
+/* fill header */
+typedef struct {
+    tvg_style_t style;
+    size_t size;
+} tvg_fill_header_t;
+
+/* line header */
+typedef struct {
+    tvg_style_t style;
+    float line_width;
+    size_t size;
+} tvg_line_header_t;
+
+/* line and fill header */
+typedef struct {
+    tvg_style_t fill_style;
+    tvg_style_t line_style;
+    float line_width;
+    size_t size;
+} tvg_line_fill_header_t;
+
+/* used to provide an input cursor over an arbitrary source */
+typedef size_t (*tvg_input_func_t)(uint8_t *data, size_t size, void *state);
+
+typedef struct {
+    /* the input source */
+    tvg_input_func_t in;
+    /* the user defined input state */
+    void *in_state;
+    /* the target pixmap */
+    twin_pixmap_t *pixmap;
+    twin_path_t *path;
+    /* the scaling used */
+    uint8_t scale;
+    /* the color encoding */
+    uint8_t color_encoding;
+    /* the coordinate range */
+    uint8_t coord_range;
+    /* the width and height of the drawing */
+    uint32_t width, height;
+    /* the size of the color table */
+    size_t colors_size;
+    /* the color table (must be freed) */
+    twin_argb32_t *colors;
+    uint32_t current_color;
+} tvg_context_t;
+
+/* the result type. TVG_SUCCESS = OK
+ * anything else, is a TVG_E_xxxx error
+ * code
+ */
+typedef int tvg_result_t;
+
+#define ERROR_CHECK_GOTO_LABEL(res, label, msg)  \
+    if (res != TVG_SUCCESS) {                    \
+        log_error("%s, error = [%d]", msg, res); \
+        goto label;                              \
+    }
+
+#define ERROR_CHECK_RETURN_CODE(res, msg)        \
+    if (res != TVG_SUCCESS) {                    \
+        log_error("%s, error = [%d]", msg, res); \
+        return res;                              \
+    }
```

Comment:
Can you follow the naming scheme of GLib? i.e., the `__return_if_fail()`, `__return_val_if_fail()`, `__return_if_reached()` and `__return_val_if_reached()` macros are intended as pre-condition assertions, to be used at the top of a public function to check that the function's arguments are acceptable.

See https://docs.gtk.org/glib/warnings.html